### ISSUE TITLE
Various issues with macro module generation

### ISSUE THREAD

**Issue by ncannasse** [HaxeFoundation/haxe#7845]
I can't get goto definition to work on IDL generated methods:

- checkout webidl https://github.com/ncannasse/webidl , 
- open webidl/sample with vscode
- compile active configuration should work 
- open Sample.hx
- F12 on p.length() - this should reach the definition within idl file

**Comment by ncannasse** [comment]
Bonus issue : if you a type in IDL file by NotExisting (or something else that is supposed to fail), the invalid type is now reported within the macro call (in TPath constructor call ??? I have no idea how that's even possible) instead of within IDL as _I think_ it was doing before.

**Comment by ncannasse** [comment]
Additional issue: the macro code adds a registerModuleDependency which is supposed to force the SampleModule to rebuild when IDL changes. But changing it (even within VSCode) does not trigger rebuild (requires restarting cache server)

**Comment by markknol** [comment]
I have simple reproducable example here [goto-settings.zip](https://github.com/HaxeFoundation/haxe/files/2893761/goto-settings.zip). A macro generates 3 fields, and creates custom position to a settings text file.

In Haxe 3.4.7 it works like this, in haxe 4.0 it doesn't goto-definition.
![goto-non-haxe-file](https://user-images.githubusercontent.com/576184/53240766-e15a4400-369f-11e9-818d-190373b11f10.gif)

**Comment by nadako** [comment]
I think the problem here is that we use `*_name_pos` for goto definition and other requests, but it is never supplied by the macro, just because it's not even in `haxe.macro.Expr` structure definitions. The `MacroApi` seems to try getting inexistant `name_pos` from the macro value and ends up using `null_pos`.

We should do some of these:
 - support `namePos` in the macro API. personally i think it's a bit confusing, and since we can't really make it a required field, it would be not enough anyway.
 - when decoding a definition from a macro, use `pos` for `name_pos` 
 - when preparing display results, fallback to `pos` if `name_pos` is `null_pos` 

PS currently if you supply a `name_pos` in your field structure from your macro (will require a `cast`) it will work :-P

**Comment by Simn** [comment]
I think I dislike 

> * when decoding a definition from a macro, use `pos` for `name_pos`

the least.

**Comment by Simn** [comment]
> Bonus issue : if you a type in IDL file by NotExisting (or something else that is supposed to fail), the invalid type is now reported within the macro call (in TPath constructor call ??? I have no idea how that's even possible) instead of within IDL as _I think_ it was doing before.

> Additional issue: the macro code adds a registerModuleDependency which is supposed to force the SampleModule to rebuild when IDL changes. But changing it (even within VSCode) does not trigger rebuild (requires restarting cache server)

You seem to be telling me that one problem is reproduced by changing the IDL file, and the other is that changing the IDL doesn't have any effect. I don't think both of these can be true at the same time.

FWIW I can't seem to reproduce either of them: Adding nonsense to point.idl gives me an error on recompilation: `point.idl:11: characters 1-19 : Unexpected Nonsense line 11`. What am I misunderstanding here?

**Comment by ncannasse** [comment]
@Simn try compiling once with good idl (+ server cache), then modify point idl and try to recompile. No error is displayed (in VSCode)

**Comment by ncannasse** [comment]
The first problem is reproducible by changing the IDL before opening vscode (no compiler cache)
The second problem requires compiler cache

**Comment by Simn** [comment]
> @Simn try compiling once with good idl (+ server cache), then modify point idl and try to recompile. No error is displayed (in VSCode)

I cannot reproduce that:

```
C:\GitHub\webidl\sample>git checkout
Your branch is up-to-date with 'origin/master'.

C:\GitHub\webidl\sample>echo class Nonsense {} > Nonsense.hx

C:\GitHub\webidl\sample>start haxe --wait 7777

C:\GitHub\webidl\sample>haxe sample.hxml --connect 7777

C:\GitHub\webidl\sample>cat point.idl Nonsense.hx >> point.idl
cat: point.idl: input file is output file

C:\GitHub\webidl\sample>cat point.idl

interface Point {
        attribute long x;
        attribute long y;
        void Point();
        void Point( long x, long y );
        [Operator="*",Ref] Point op_add( [Const,Ref] Point p );
        double length();
};

class Nonsense {}

C:\GitHub\webidl\sample>haxe sample.hxml --connect 7777
point.idl:11: characters 1-20 : Unexpected Nonsense line 11
Sample.hx:1: characters 8-20 : Module SampleModule does not define type SampleModule
```

I'll go ahead and assume that this has been fixed. Please reopen if you can still reproduce it.

**Comment by ncannasse** [comment]
New (more precise) instructions for second bug:

Add the following line to the class Point in sample.idl:

```NotExisting foo();```

Compile. 
Error is reported within Module.hx (TPath call) instead of the inserted position (within idl file). Most likely a position mistake (using the runtime one)

I confirm original issue + third issues seems to be fixed. Only second issue remains.

**Comment by Simn** [comment]
I don't really understand how you expect this to use any other position. You use `ret : makeType(ret)`, so you end up with a `ComplexType` and these don't have a position on the macro-end.

We might be missing API here, but this doesn't look like a regression.

**Comment by ncannasse** [comment]
It is a regression, it used to give the error at the class_field position. But since we introduced type_hint in parser which assigns a pos to each complex type, we cannot define these in macros.

**Comment by ncannasse** [comment]
I think this requires exposing TypeHint to the macro AST, as an abstact with a from/to ComplexType for maximum backward compatibility.

**Comment by Simn** [comment]
Yes that's what I meant by "missing API". This might be a bit messy with variance and that fun `Const` flag on structures.

**Comment by Simn** [comment]
I'm not sure what to do with this. Introducing these compatibility abstracts is likely going to cause a lot of friction which I don't want to deal with right now. I would prefer to not touch this for 4.0 and then look at it again later.

**Comment by Simn** [comment]
I'm moving this to 4.1. While I acknowledge that having inaccurate error positions in such cases isn't ideal, it's not urgent enough to warrant larger reviews of the macro API at this point. I can't see a way to address this without friction, so I'd appreciate if you could just find a workaround for this for now.
